### PR TITLE
Refactor encryption-related code

### PR DESCRIPTION
**Pull request by dennwc**
* Change the type of AccessPermissions to bitmask instead of struct with bools - this will preserve unsupported permission bits.
* Split common field fo Encrypt dictionary and field related to Standard security handler.
* Make a proper constructor for PdfCrypt for writing a new document.
* Unexport methods related to crypt filters, make CryptFilter an interface.
* Unexport all Encrypt dictionary fields, fix other code to access them properly.

### CODE REVIEW

File: pdf/core/crypt.go [hunk 1]
```diff
@@ -6,168 +6,313 @@
...
+	if length, ok := d.Get("Length").(*PdfObjectInteger); ok {
+		cf.Length = int(*length)
+	}
+	return nil
 }
 
-// CryptFilters is a map of crypt filter name and underlying CryptFilter info.
-// TODO (v3): Unexport.
-type CryptFilters map[string]CryptFilter
+func (crypt *PdfCrypt) newEncyptDict() *PdfObjectDictionary {
```
Comment (gunnsth) [inline pdf/core/crypt.go]: typo Encypt/Encrypt

File: pdf/core/security/auth.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package security
```
Comment (gunnsth) [inline pdf/core/security/auth.go]: https://github.com/unidoc/unidoc/wiki/UniDoc-Developer-Guide#8-license-header-in-each-go-file

File: pdf/core/security/auth.go [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+// AuthEvent is an event type that triggers authentication.
+type AuthEvent string
+
+const (
+	EventDocOpen = AuthEvent("DocOpen") // document open
```
Comment (gunnsth) [inline pdf/core/security/auth.go]: godoc comments for all exported symbols

File: pdf/core/security/crypt/filter_aesv3.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	if d.Length != 0 && d.Length != 32 {
+		return nil, fmt.Errorf("Invalid AESV3 crypt filter length (%d)", d.Length)
+	}
+	return filterAESV3{}, nil
+}
+
+// filterAES implements a generic AES encryption and decryption algorithm used by AESV2 and AESV3 filter methods.
+type filterAES struct{}
+
+func (filterAES) EncryptBytes(buf []byte, okey []byte) ([]byte, error) {
```
Comment (gunnsth) [inline pdf/core/security/crypt/filter_aesv3.go]: godoc on exported symbol

File: pdf/core/security/crypt/filter_aesv3.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	mode := cipher.NewCBCEncrypter(ciph, iv)
+	mode.CryptBlocks(ciphertext[block:], buf)
+
+	buf = ciphertext
+	common.Log.Trace("to (%d): % x", len(buf), buf)
+
+	return buf, nil
+}
+
+func (filterAES) DecryptBytes(buf []byte, okey []byte) ([]byte, error) {
```
Comment (gunnsth) [inline pdf/core/security/crypt/filter_aesv3.go]: godoc on exported symbol.
// DecryptBytes implements interface security.Filter.

File: pdf/core/security/crypt/filters.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+func getFilter(name string) (filterFunc, error) {
+	f := filterMethods[string(name)]
+	if f == nil {
+		return nil, fmt.Errorf("unsupported crypt filter: %q", name)
+	}
+	return f, nil
+}
+
+// Filter is a common interface for crypt filter methods.
+type Filter interface {
```
Comment (gunnsth) [inline pdf/core/security/crypt/filters.go]: Place towards top of file.

File: pdf/core/security/standard_r4.go [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+	"bytes"
+	"crypto/md5"
+	"crypto/rand"
+	"crypto/rc4"
+	"errors"
+
+	"github.com/unidoc/unidoc/common"
+)
+
+var _ StdHandler = stdHandlerR4{}
```
Comment (gunnsth) [inline pdf/core/security/standard_r4.go]: Why is this needed?

File: pdf/core/security/standard_r4.go [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+const padding = "\x28\xBF\x4E\x5E\x4E\x75\x8A\x41\x64\x00\x4E\x56\xFF" +
+	"\xFA\x01\x08\x2E\x2E\x00\xB6\xD0\x68\x3E\x80\x2F\x0C" +
+	"\xA9\xFE\x64\x53\x69\x7A"
+
+// NewHandlerR4 creates a new standard security handler for R<=4.
+func NewHandlerR4(id0 string, length int) StdHandler {
+	return stdHandlerR4{ID0: id0, Length: length}
+}
+
+type stdHandlerR4 struct {
```
Comment (gunnsth) [inline pdf/core/security/standard_r4.go]: A comment would be nice even if not exported.

File: pdf/core/security/standard_r4.go [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+	}
+
+	if !bytes.Equal(uGen, uDoc) {
+		return nil, nil
+	}
+	return ekey, nil
+}
+
+// alg7 authenticates the owner password and returns the document encryption key.
+//// It returns an nil key in case authentication failed.
```
Comment (gunnsth) [inline pdf/core/security/standard_r4.go]: Any reason for double comment //// ?

File: pdf/core/security/standard_r4.go [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+	if err != nil {
+		common.Log.Debug("ERROR: Error generating O for encryption (%s)", err)
+		return nil, err
+	}
+	d.U = U
+	common.Log.Trace("gen U: % x", U)
+	return ekey, nil
+}
+
+func (sh stdHandlerR4) Authenticate(d *StdEncryptDict, pass []byte) ([]byte, Permissions, error) {
```
Comment (gunnsth) [inline pdf/core/security/standard_r4.go]: Would be nice to document, just knowing what interface it implements is useful.

File: pdf/core/security/standard_r6.go [hunk 11]
```diff
@@ -0,0 +1,466 @@
...
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/binary"
+	"errors"
+	"hash"
+	"io"
+	"math"
+)
+
+var _ StdHandler = stdHandlerR6{}
```
Comment (gunnsth) [inline pdf/core/security/standard_r6.go]: again not sure why this is needed?

File: pdf/model/writer.go [hunk 12]
```diff
@@ -11,19 +11,18 @@ package model
...
-	"crypto/rand"
 	"encoding/binary"
 	"errors"
 	"fmt"
 	"io"
-	"math"
 	"strings"
-	"time"
+
+	"github.com/unidoc/unidoc/pdf/core/security/crypt"
```
Comment (gunnsth) [inline pdf/model/writer.go]: extra line

File: pdf/core/security/crypt/filter_aesv3.go (hunk 4 above)
Comment (dennwc) [reply pdf/core/security/crypt/filter_aesv3.go]: It can be considered an unexported symbol since it's defined on unexported type. But I added a godoc anyway.

File: pdf/core/security/standard_r4.go (hunk 7 above)
Comment (dennwc) [reply pdf/core/security/standard_r4.go]: This is a compile-time assertion for "stdHandlerR4 implements StdHandler". It is also useful to see what interface this type implements. I can remove them, but I personally find them rather useful.

**Comment by dennwc** [comment]
@gunnsth Updated, please take a look. Also, opened a PR to examples repo: https://github.com/unidoc/unidoc-examples/pull/19.

Answering the question, it's not possible to move `core/security` to `internal` because it defines the `Permissions` type, and this type cannot be defined in `core` because it will create an import cycle.
